onnxruntime/python/tools/quantization/calibrate_user.py

```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python
+# coding: utf-8
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft, Intel Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import os
+import sys
+import numpy as np
+import re
+import subprocess
+import json
+
+import abc
+
+import onnx
+import onnxruntime
+from onnx import helper, TensorProto, numpy_helper
+from quantize import quantize, QuantizationMode
+from calibrate import calibrate
+
+#user-implement preprocess func
+from data_preprocess import preprocess_func
+
+
+class CalibrationDataReaderInterface(metaclass=abc.ABCMeta):
+    @classmethod
+    def __subclasshook__(cls,subclass):
+        return (hasattr(subclass,'get_next') and callable(subclass.get_next) or NotImplemented)
+
+    @abc.abstractmethod
+    def get_next(self) -> dict:
+        """generate the input data dict for ONNXinferenceSession run"""
+        raise NotImplementedError
+
+class CalibrationDataReader(CalibrationDataReaderInterface):
+    def __init__(self,calibration_image_folder,augmented_model_path='augmented_model.onnx'): 
+        self.image_folder = calibration_image_folder
+        self.augmented_model_path = augmented_model_path
+        self.preprocess_flag = True
+        self.enum_data_dicts = []
+        self.datasize = 0
+
+    def get_next(self):
+        if self.preprocess_flag:
+            self.preprocess_flag = False
+            session = onnxruntime.InferenceSession(self.augmented_model_path, None)
+            (_,height,width,_) = session.get_inputs()[0].shape
+            nhwc_data_list = preprocess_func(self.image_folder,height,width,size_limit = 0)
+            input_name = session.get_inputs()[0].name
+            self.datasize = len(nhwc_data_list)           
+            self.enum_data_dicts = iter([{input_name:nhwc_data_list[i]} for i in range(self.datasize)])
+        return next(self.enum_data_dicts,None)
+
+def main():
+    model_path = 'path/to/model.onnx'
+    calibration_dataset_path = 'path/to/calibration_data_set'
```

Comment:
it'd be better to make it a workable example. we can change the E2E_example_model to work with this new api: [https://github.com/microsoft/onnxruntime/tree/master/onnxruntime/python/tools/quantization/E2E_example_model](https://github.com/microsoft/onnxruntime/tree/master/onnxruntime/python/tools/quantization/E2E_example_model)